url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/DeviceGraphicsDisplay/LaneGraphicsDisplay.java

```diff
@@ -0,0 +1,510 @@
+package DeviceGraphicsDisplay;
+
+import java.awt.Graphics2D;
+import java.util.ArrayList;
+
+import javax.swing.JComponent;
+
+import Networking.Client;
+import Networking.Request;
+import Utils.Constants;
+import Utils.Location;
+import Utils.PartData;
+import factory.PartType;
+
+/**
+ * This class contains the graphics display components for a lane.
+ * 
+ * @author Shalynn Ho
+ * 
+ */
+public class LaneGraphicsDisplay extends DeviceGraphicsDisplay {
+	// max number of parts that can be on a Lane
+	private static final int MAX_PARTS = Constants.LANE_LENGTH
+			/ (Constants.PART_WIDTH + Constants.PART_PADDING / 2);
+	// number of lines on the lane
+	private static final int NUMLINES = 1 + Constants.LANE_LENGTH
+			/ (4 * Constants.PART_WIDTH);
+	// space between each line
+	private static final int LINESPACE = Constants.LANE_LENGTH / NUMLINES;
+	// width of lane lines
+	private static final int LINE_WIDTH = 3;
+
+	// stores the parts on the lane
+	private final ArrayList<PartGraphicsDisplay> partsOnLane;
+
+	// start location of parts on this lane
+	private final Location partStartLoc;
+
+	// location of a part jam on the lane
+	private Location jamLoc;
+	private int jamLocX = 0;
+
+	// array list of locations of the lane lines
+	private ArrayList<Location> laneLines;
+
+	// the ID of this Lane
+	private final int laneID;
+
+	// the amplitude of this lane
+	private int speed = 2;
+	private int amplitude = 2;
+
+	// true if Lane is on
+	private boolean laneOn = true;
+
+	// use to make sure only 1 message is sent to agent for each part that
+	// reaches end of lane
+	private boolean receivePartDoneSent = false;
+	private boolean purgeDoneSent = false;
+
+	// state of the parts on lane
+	private boolean partAtLaneEnd = false;
+	private boolean purging = false;
+	private boolean jammed = false;
+	private boolean unjamming = false;
+	private int jamSeq = 0;
+
+	/**
+	 * LGD constructor
+	 * 
+	 * @param lm
+	 *            - the lane manager (client)
+	 * @param lid
+	 *            - lane ID
+	 */
+	public LaneGraphicsDisplay(Client c, int lid) {
+		client = c;
+		laneID = lid;
+
+		partsOnLane = new ArrayList<PartGraphicsDisplay>();
+		// generate lane start location
+		location = new Location(Constants.LANE_END_X, 53 + laneID * 75);
+
+		jamLoc = new Location(Constants.LANE_END_X + Constants.LANE_LENGTH / 2,
+				location.getY());
+
+		// for reference only
+		partStartLoc = new Location(Constants.LANE_BEG_X, location.getY()
+				+ Constants.PART_WIDTH / 2 - Constants.PART_OFFSET);
+
+		resetLaneLineLocs();
+	}
+
+	/**
+	 * Animates lane movement and sets location of parts moving down lane
+	 * 
+	 * @param c
+	 *            - component on which this is drawn
+	 * @param g
+	 *            - the graphics component on which this draws
+	 */
+	@Override
+	public void draw(JComponent c, Graphics2D g) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.